### PR TITLE
`build`: generate hex dump of server assets during build

### PR DESCRIPTION
**Pull request by ochafik**
Currently one needs to manually regenerate asset files in `examples/server`, which:
* is easy to forget
* ~~may cause a security risk~~ (reviewers are unlikely to unescape the code - don't wanna be the next target after [xz](https://en.wikipedia.org/wiki/XZ_Utils_backdoor)) (*edit*: already mitigated by https://github.com/ggerganov/llama.cpp/pull/6409 as pointed out below)

`od` is used in `Makefile` (unlike `xxd`, it should be available on ~all systems, incl. on Windows w/ `w64devkit` - thanks to busybox), while `cmake` and `zig` builds cast their own hex for portability.

### CODE REVIEW

**Comment by phymbert** [comment]
It has been addressed the xz weekend here:
- #6409

But I am happy that finally the code is removed, so looks good

**Comment by ngxson** [comment]
Please pay attention that the idea may look simple on the surface, but it's get messy when taking into account that we're not compiling for single target.
- Some linux distro (docker/CI) may not have `xxd` by default. You may need to add something like `apt install xxd` to every dockerfile and compile scripts in the project. This will also create a new dependency in compilation phase. In fact, a while ago, I success in making a script that does the same thing with `xxd` but using `od` instead. The reason was because `od` is included by default in most linux distros.
- Things get messy when talking about Windows and MacOS. In fact, I have 0 knowledge how to compile on these targets, so I gave up.

**Comment by ochafik** [comment]
@phymbert great to read you've already mitigated the issue, I'd missed that commit in the flow.

> Things get messy when talking about Windows and MacOS. In fact, I have 0 knowledge how to compile on these targets, so I gave up.

@ngxson actually I've just hit a snag by using `-n` that my BSD xxd likes (on MacOS), and GNU xxd has no clue about (on CI) - fixed.

Re/ Windows, I realize I've assumed people build from within WSL (or cross-build from Linux), but would need confirmation.

In the absolute worst case we could write our own minimalistic xxd, I suppose.

**Comment by phymbert** [comment]
> Re/ Windows, I realize I've assumed people build from within WSL (or cross-build from Linux), but would need confirmation.

We cannot assume unfortunately, I have the feeling most windows developper are using Visual Studio.

> In the absolute worst case we could write our own minimalistic xxd, I suppose.

I would prefer to have a classical npm build then. Also what about:
- https://stackoverflow.com/questions/4158900/embedding-resources-in-executable-using-gcc

**Comment by ochafik** [comment]
> I would prefer to have a classical npm build then.

@phymbert possibly inevitable in the long run, would unlock new horizons...

> Also what about: https://stackoverflow.com/questions/11813271/embed-resources-eg-shader-code-images-into-executable-library-with-cmake

Hah, nifty! Might need some contortions to get something similar to work w/ MSVC, but alternatively... looks like we could do it in cmake: https://stackoverflow.com/questions/11813271/embed-resources-eg-shader-code-images-into-executable-library-with-cmake

Which means out of the 3 documented ways to build on Windows:

- (MS) `make` + `w64devkit`: the latter ships w/ busybox, which seems to have `xxd`
- `cmake`: can do the HEX in CMakeLists.txt
- `zig`: same (TBC, I'm not sure how much I like the language so far :-D)

(so xxd in Makefile for all platforms, and adhoc hexing in the other build scripts 🤞)

**Comment by ngxson** [comment]
Good job! The read file as HEX trick in cmake seem to resolve the problem on both mac & windows.

For `make`, there're users still using `make` on linux, and `xxd` may not be available by default, at least it's not installed on my fedora installation and my ubuntu distrobox. Of course I can simply `apt install`, but I'm a bit doubt that this change may break compilation for someone else.

File: examples/server/CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,26 @@
...
+    set(input "${CMAKE_CURRENT_SOURCE_DIR}/public/${asset}")
+    set(output "${CMAKE_CURRENT_BINARY_DIR}/${asset}.hpp")
+
+    string(REGEX REPLACE "\\.| |-" "_" name "${asset}")
+    file(READ "${input}" data HEX)
+    string(LENGTH ${data} hex_len)
+    math(EXPR len "${hex_len} / 2")
+    string(REGEX REPLACE "([0-9a-f][0-9a-f])" "0x\\1," data ${data})
+    file(WRITE "${output}" "unsigned char ${name}[] = {${data}};\nunsigned int ${name}_len = ${len};\n")
+endforeach()
```
Comment (cebtenzzre) [inline examples/server/CMakeLists.txt:18]: Doing this at configure time seems wrong - if the files change between running `cmake` (which shouldn't need to be done frequently) and running `make` or `cmake --build`, they will not be updated.

**Comment by cebtenzzre** [comment]
Nomic's kompute fork builds xxd from source (it's a single .c source file) unconditionally: https://github.com/nomic-ai/kompute/blob/d1e3b0953cf66acc94b2e29693e221427b2c1f3f/CMakeLists.txt#L187

**Comment by ochafik** [comment]
> Nomic's Kompute fork builds xxd from source (it's a single .c source file) unconditionally: https://github.com/nomic-ai/kompute/blob/d1e3b0953cf66acc94b2e29693e221427b2c1f3f/CMakeLists.txt#L187

@cebtenzzre hah great, so we kinda always have xxd after all... (assuming one does `git clone --recursive`, which I didn't)

>> In the absolute worst case we could write our own minimalistic xxd, I suppose.

> This does not sound really awful - we can give it a try if we encounter issues with the proposed PR

@ggerganov @ngxson I've switched the Makefile to using `od` & `wc` (should work on Windows w/ w64devkit through busybox 🤞). None of the 3 hex-dumping methods are exactly pretty, but they're relatively short (building `xxd` would probably take the same amount of build script lines, or more b/c of two-phased builds).

**Review by ggerganov — APPROVED** [review]

**Review by ngxson — APPROVED** [review]

**Comment by sorasoras** [comment]
@ochafik 
I think this kind of break rocm for cmake in windows.

[debugbuild.txt](https://github.com/ggerganov/llama.cpp/files/15062347/debugbuild.txt)

**Comment by ochafik** [comment]
> I think this kind of break rocm for cmake in windows.
> [debugbuild.txt](https://github.com/ggerganov/llama.cpp/files/15062347/debugbuild.txt)

@sorasoras glancing through it, think it might related to https://github.com/ggerganov/llama.cpp/pull/6797 @jart FYI

**Pull request merged**